### PR TITLE
Replace outdated libmp4 with ffmpeg demuxer for FAAD2

### PR DESCRIPTION
**Pull request by daschuer**
This is a solution for m4a decoding that replaces libmp4 banned by Debian with ffmpeg. It keeps the FAAD2 decoder to not introduce an offset that invalidated the track notations. 

This allows us to establish an offset detection based on the silence detection that will be introduced along with 2.3. 

Without this from Ubuntu Focal user have no chance to generate the silence detection data with FAAD2 required for the upcoming offset detection.  

This was original discussed here: https://github.com/mixxxdj/mixxx/pull/3050

### CODE REVIEW

**Comment by daschuer** [comment]
@uklotzde: Now the sound of this version is exactly 2112 frames earlier than the original version. 
I cant explain this, because for my understanding I feet  the same sample blocks to libFaad like before. 
Maybe this is related to the MP4ESConfigBuffer data or there is a subtle bug. 
Do you have an idea?

**Comment by daschuer** [comment]
MP4ESConfigBuffer is the same. There is a stream layout issue and SoundsourceFFmpeg takes over. I think that is the reason for the 2112 frames offset.

**Comment by daschuer** [comment]
I need to configure <= two channels than It should work. Stay tuned.

**Comment by daschuer** [comment]
Done :-) 
Now libfaad receives the same blocks of data as in the libmp4 version. 
So there should be no difference. 

Obviously there is a lot of room for improvements. However I like to keep the implementation close to the original, to not introduce any difference. Given that this is only a temporary solution I am fine with it.

**Comment by daschuer** [comment]
Should we also backport it to 2.2 to bring back m4a audio in our stable Ubuntu Focal and derivates?

File: src/sources/soundsourcem4a.cpp [hunk 1]
```diff
@@ -54,96 +70,73 @@ constexpr MP4Duration kDefaultFramesPerSampleBlock = 1024;
...
-    return MP4GetNumberOfTracks(hFile, nullptr, 0);
+inline bool isValidMediaDataName(const char* mediaDataName) {
+    return (nullptr != mediaDataName) &&
+            (0 == strcasecmp(mediaDataName, "mp4a"));
 }
 
-inline bool isValidTrackType(const char* trackType) {
-    return (nullptr != trackType) &&
-            MP4_IS_AUDIO_TRACK_TYPE(trackType);
+QString formatErrorMessage(int errnum) {
```
Comment (uklotzde) [inline src/sources/soundsourcem4a.cpp:78]: I have already fixed this for SoundSourceFFmpeg

File: src/sources/soundsourcem4a.cpp [hunk 2]
```diff
@@ -475,28 +495,21 @@ void SoundSourceM4A::close() {
...
     m_inputBuffer.clear();
-    if (m_pMP4ESConfigBuffer) {
-        free(m_pMP4ESConfigBuffer);
-        m_pMP4ESConfigBuffer = nullptr;
-    }
-    if (MP4_INVALID_FILE_HANDLE != m_hFile) {
-        MP4Close(m_hFile);
-        m_hFile = MP4_INVALID_FILE_HANDLE;
-    }
+    m_pAvInputFormatContext.close();
```
Comment (uklotzde) [inline src/sources/soundsourcem4a.cpp:498]: also reset m_pAvStream

File: src/sources/soundsourcem4a.cpp [hunk 3]
```diff
@@ -506,6 +519,58 @@ void SoundSourceM4A::restartDecoding(MP4SampleId sampleBlockId) {
...
+    } else if (packet.size <= static_cast<int>(*pNumBytes)) {
+        memcpy(*ppBytes, packet.data, packet.size);
+        m_nextSampleBlockId = sampleBlockId + 1;
+        DEBUG_ASSERT(packet.pos == m_pAvStream->index_entries[sampleBlockId - 1].pos);
+    } else {
+        error = AVERROR(ENOMEM);
+    }
+    if (sampleBlockId == kSampleBlockIdMin) {
+        // first block contains extra data that make libfaad struggle.
+        *pNumBytes = 6;
```
Comment (uklotzde) [inline src/sources/soundsourcem4a.cpp:566]: Why 6? And without any safety checks that packet.size is indeed >= 6 as expected?

File: src/sources/soundsourcem4a.cpp [hunk 4]
```diff
@@ -631,7 +694,9 @@ ReadableSampleFrames SoundSourceM4A::readSampleFramesClamped(
                     m_inputBufferLength = inputBufferLength;
                     m_inputBufferOffset = 0;
                     // Skip ADTS header if we have a decoder specific ES config
-                    if (m_pMP4ESConfigBuffer &&
+                    AVCodecParameters* par = m_pAvStream->codecpar;
+                    DEBUG_ASSERT(par != nullptr);
+                    if (par->extradata && par->extradata_size > 0 &&
```
Comment (uklotzde) [inline src/sources/soundsourcem4a.cpp:699]: Your usage of `!= nullptr` is inconsistent. Here you omit it whereas in other situations you add it.

File: src/sources/soundsourcem4a.cpp [hunk 5]
```diff
@@ -506,6 +519,58 @@ void SoundSourceM4A::restartDecoding(MP4SampleId sampleBlockId) {
...
+
+    m_nextSampleBlockId = kSampleBlockIdInvalid;
+
+    error = av_read_frame(m_pAvInputFormatContext, &packet);
+    if (error < 0) {
+        av_make_error_string(error_string, AV_ERROR_MAX_STRING_SIZE, error);
+        qWarning() << "Could not read frame:" << error_string;
+    } else if (packet.size <= static_cast<int>(*pNumBytes)) {
+        memcpy(*ppBytes, packet.data, packet.size);
+        m_nextSampleBlockId = sampleBlockId + 1;
```
Comment (uklotzde) [inline src/sources/soundsourcem4a.cpp:559]: Do you really need m_nextSampleBlockId for keeping track? I guess m_curSampleBlockId could be sufficient.

File: src/sources/soundsourcem4a.cpp [hunk 6]
```diff
@@ -60,6 +60,16 @@ constexpr SINT kNumberOfPrefetchFrames = 2112;
...
 
+// According to various references DecoderConfigDescriptor.bufferSizeDB
+// is a 24-bit unsigned integer value.
+// MP4 atom:
+//   trak.mdia.minf.stbl.stsd.*.esds.decConfigDescr.bufferSizeDB
+// References:
+//   https://github.com/sannies/mp4parser/blob/master/isoparser/src/main/java/org/mp4parser/boxes/iso14496/part1/objectdescriptors/DecoderConfigDescriptor.java
+//   http://mutagen-specs.readthedocs.io/en/latest/mp4/
+//   http://perso.telecom-paristech.fr/~dufourd/mpeg-4/tools.html
+constexpr u_int32_t kMaxSampleBlockInputSizeLimit = (u_int32_t(1) << 24) - 1;
```
Comment (Holzhaus) [inline src/sources/soundsourcem4a.cpp:71]: What is the difference between `u_int32_t` and `uint32_t` (used below in line 525)? Let's use a single naming. (maybe `quint2`?)

File: src/sources/soundsourcem4a.cpp (hunk 3 above)
Comment (daschuer) [reply src/sources/soundsourcem4a.cpp:566]: Maybe you have some insights. 
With libmp4 it is 
33 0 3 64 104 28
With ffmpeg
1 33 0 3 64 104 28 64 8 0 0 0 0 0 0 0 0 0 0 0 10 198 140 77 157 16 142 37 233 254

Comment (daschuer) [reply src/sources/soundsourcem4a.cpp:566]: Oh I have just noticed that his can be quite different.

**Comment by daschuer** [comment]
OK, I have discovered some odd cases where this behaves differently the libmp4 version. 
It looks like that the demuxer algorithm is manly responsible to create offsets.   
If we start to code around this we fail the original requirement to have no offset compared to limp4 + faad. 

I will close this now.

**Pull request closed without merging**